mri/reconstructors/self_calibrating.py

```diff
@@ -0,0 +1,180 @@
+                 smaps_gridding_method='linear', n_jobs=1, verbose=0):
+        if linear_op is None:
+            linear_op = WaveletN(
+                wavelet_name="sym8",
+                nb_scale=3,
+                dim=len(fourier_op.shape),
+                n_coils=1,
+                verbose=bool(verbose >= 30),
+            )
+        # Ensure that we are in right multichannel config
+        if linear_op.n_coils != 1:
+            raise ValueError("The value of n_coils for linear operation must "
+                             "be 1 for Self-Calibrating reconstruction!")
+        if gradient_method == 'analysis':
+            grad_class = GradSelfCalibrationAnalysis
+        elif gradient_method == 'synthesis':
+            grad_class = GradSelfCalibrationSynthesis
+        super(SelfCalibrationReconstructor, self).__init__(
+            fourier_op=fourier_op,
+            linear_op=linear_op,
+            prox_op=prox_op,
+            mu=mu,
+            gradient_method=gradient_method,
+            grad_class=grad_class,
+            lipschitz_cst=lipschitz_cst,
+            lips_calc_max_iter=lips_calc_max_iter,
+            num_check_lips=num_check_lips,
+            init_gradient_op=False,
+            verbose=verbose,
+        )
+        self.smaps_gridding_method = smaps_gridding_method
+        self.smaps_extraction_mode = smaps_extraction_mode
+        if type(kspace_portion) == float:
+            self.kspace_portion = (kspace_portion,) * \
+                                  self.fourier_op.samples.shape[-1]
+        else:
+            self.kspace_portion = kspace_portion
+        if len(self.kspace_portion) != self.fourier_op.samples.shape[-1]:
+            raise ValueError("The k-space portion size used to estimate the" +
+                             "sensitivity information is not aligned with" +
+                             " the input dimension")
+        self.n_jobs = n_jobs
+
+    def reconstruct(self, kspace_data, optimization_alg='pogm', x_init=None,
+                    num_iterations=100, reinit_grad_op=False, **kwargs):
+        """ This method calculates operator transform.
+        Parameters
+        ----------
+        kspace_data: np.ndarray
+            the acquired value in the Fourier domain.
+            this is y in above equation.
+        optimization_alg: str (optional, default 'pogm')
+            Type of optimization algorithm to use, 'pogm' | 'fista' |
+            'condatvu'
+        x_init: np.ndarray (optional, default None)
+            input initial guess image for reconstruction
+        num_iterations: int (optional, default 100)
+            number of iterations of algorithm
+        """
+        if self.fourier_op.n_coils != kspace_data.shape[0]:
+            raise ValueError("The provided number of coil (n_coils) do not "
+                             "match the data itself")
+        # Extract Sensitivity maps and initialize gradient
+        Smaps, _ = get_Smaps(
```

Comment:
`recompute_smaps` should be added here and the Lispschitz constant should be calculated again only if Smaps have been recomputed

Reply:
In my opinion it is same as reinit_grad_op. The only corner case is having same gradient operator, but reinitialize the lipchitz constant, I am not sure if this is a valid use case

Reply:
Nope it is not, you could provide an a priori sensitivity maps and just want to update the gradient lipschitz constante, in which case you want to `recompute_lipschitz`, however `recompute_smaps` extracts the senitivity maps from the input data and recompute the Lipschitz constant